Template Part Block: Use buttons for area selection

creativecoder opened the pull request:
## Description

Resolves #29295

## How has this been tested?

1. Open the site editor
2. Select a template part block, such as the header (the list view is an easy way to do this).
3. Open the Advanced section of the sidebar options and choose the area.

## Screenshots <!-- if applicable -->

Template Part Block > Advanced settings in sidebar

<img width="273" alt="image" src="https://user-images.githubusercontent.com/1699996/111818830-a2624100-88ad-11eb-8270-0cfda68c191c.png">

<img width="410" alt="image" src="https://user-images.githubusercontent.com/1699996/111832735-2a047b80-88bf-11eb-8d89-73f61ed66e7a.png">



## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

creativecoder committed:
Template Part Block: Use buttons for area selection

jameskoster committed:
Update general area icon and description

jameskoster committed:
formatting

jameskoster commented:
Looking good! :) 

I updated the icon / description for the general area.

<img width="869" alt="Screenshot 2021-03-19 at 09 24 33" src="https://user-images.githubusercontent.com/846565/111760502-b3ec1e80-8896-11eb-90a2-9383ce66edd7.png">

The width of these tooltips makes them a little awkward to read. Is it possible to wrap them?

creativecoder committed:
Format template part area descriptions

creativecoder commented:
> I updated the icon / description for the general area.

Thank you!

> The width of these tooltips makes them a little awkward to read. Is it possible to wrap them?

Yes! Now updated.

I'm a little unsure about the screen reader accessibility of using tooltips to describe the template part areas. I modified the Tooltip component to be able to receive an id and be the target of aria-describedby (declared on the buttons)--this seems to help when testing with VoiceOver, but I'd want a more experienced opinion before merging this.

creativecoder committed:
Removes duplicate prop

david-szabo97 commented on packages/components/src/tooltip/index.js:
```diff
@@ -87,7 +91,14 @@ const emitToChild = ( children, eventName, event ) => {
+	position,
+	text,
+	shortcut,
+	...popoverProps
```
Hmmmm 🤔 This seems to be a rather big modification. I wonder if we should use an explicit `popoverProps` property instead of `...` rest. We need to update the README too.

As a last thought, maybe we should move this to a separate PR?

david-szabo97 commented:
It's working as expected. My only concern is the modification of the `Tooltip` component. The changes make sense but it feels it should be its own PR. Looking forward to others' thoughts.

https://user-images.githubusercontent.com/2256104/111965470-0ecf8100-8af6-11eb-96cb-558d0531462d.mov

jameskoster commented:
+1 to moving the a11y discussion to a separate issue since it will effectively relate to all tooltip usage across the app.

david-szabo97 commented:
@creativecoder Could you break this into multiple PRs?

We should remove the tooltip changes from this PR and create a separate PR for just the tooltip. Then we can create a follow-up PR to add the accessibility parts once we have the tooltip PR merged. You can use the tooltip PR as a base for the follow-up PR so we can show a use-case as well in the tooltip PR.

So basically we will have 3 PRs in the end:
1. This PR which introduces the buttons for area selection (without a11y in mind)
2. Tooltip PR which adds the new prop to the `Tooltip` component so we can passthrough props to the `Popover` component
3. A accessibility follow-up PR to the 1. one, using 2. as the base branch

creativecoder commented:
> Could you break this into multiple PRs?

Yes! Will work on that soon. I'd like to leave the className passthrough (Tooltip prop that's passed to the Popover component) so we can control the tooltip width. That seems less invasive, and therefore okay? I'll be sure to update the component readme, accordingly.

david-szabo97 commented:
> I'd like to leave the className passthrough (Tooltip prop that's passed to the Popover component) so we can control the tooltip width. That seems less invasive, and therefore okay?

I would be still concerned about the PR's responsibility. In my opinion, it's better to avoid making core changes in a mostly Site Editor related PR. Smaller PRs and single responsibility PRs are easier to revert and discuss.

Addison-Stavlo commented on packages/block-library/src/template-part/edit/advanced-controls.js:
```diff
@@ -53,13 +78,41 @@ export function TemplatePartAdvancedControls( {
-						value={ area }
-						onChange={ setArea }
-					/>
+					<BaseControl
```
I wonder if `RadioGroup`/`Radio` [components](https://developer.wordpress.org/block-editor/reference-guides/components/radio-group/) would be able to fit well and be adaptable for this case?  It seems pretty similar visually and behaviorally to what we are going for.

david-szabo97 replied:
+1 for RadioGroup and Radio components. They seem to fit this particular situation.

creativecoder replied:
Thanks for the idea, I will take a look.

Addison-Stavlo commented on packages/block-library/src/template-part/edit/advanced-controls.js:
```diff
@@ -2,19 +2,44 @@
-	{ label: __( 'Header' ), value: 'header' },
-	{ label: __( 'Footer' ), value: 'footer' },
 	{
+		description: __(
+			'The Header template defines a page area that typically contains a title, logo, and main navigation.'
+		),
+		icon: header,
+		label: __( 'Header' ),
+		value: 'header',
```
Not something that needs changed in this PR, but we should centralize these descriptions/definitions so we don't need to re-define them in a handful of places.  I'm trying that as part of https://github.com/WordPress/gutenberg/pull/30395, so a follow up once these both land would be to update this to get that data from the core editor store.

Addison-Stavlo reviewed (commented):
Just some minor thoughts, its lookin pretty good so far. 😁

jameskoster reviewed (approved):
Approving the design. If the width has to be fixed in a follow-up PR, so be it :D

david-szabo97 commented:
👋 Would love to see this merged 😄 Could you rebase it? Happy to approve once tests are passing and rebased. We can create a follow-up PR for the [a11y](https://github.com/WordPress/gutenberg/pull/30005#discussion_r607144676) comment.

creativecoder commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/45764, which uses the newer ToggleGroupControl component.
